inc/project.class.php

```diff
@@ -332,46 +332,106 @@ function cleanDBonPurge() {
+      global $DB;
 
-      $join = '';
+      Toolbox::deprecated('Use getVisibilityCriteria');
 
-      if (!Session::haveRight("project", Project::READALL)) {
-         $join .= " LEFT JOIN `glpi_projectteams`
-                     ON (`glpi_projectteams`.`projects_id` = `glpi_projects`.`id`) ";
-      }
+      //get and clean criteria
+      $criteria = self::getVisibilityCriteria();
+      unset($criteria['WHERE']);
+      $criteria['FROM'] = self::getTable();
 
-      return $join;
+      $it = new \DBmysqlIterator(null);
+      $it->execute($criteria);
+      $sql = $it->getSql();
+      $sql = str_replace(
+         'SELECT * FROM '.$DB->quoteName(self::getTable()).' ',
+         '',
+         $sql
+      );
+      return $sql;
    }
 
    /**
     * Return visibility to add to SQL
     *
+    * @deprecated 9.4.0
+    *
     * @return string joins to add
     **/
    static function addVisibility() {
+      Toolbox::deprecated('Use getVisibilityCriteria');
 
-      $condition = '';
-      if (!Session::haveRight("project", Project::READALL)) {
-         $teamtable = 'glpi_projectteams';
-         $condition .= "AND (`glpi_projects`.users_id = '" . Session::getLoginUserID() . "'
-                               OR (`$teamtable`.`itemtype` = 'User'
-                                   AND `$teamtable`.`items_id` = '" . Session::getLoginUserID() . "')";
-         if (count($_SESSION['glpigroups'])) {
-            $condition .= " OR (`glpi_projects`.`groups_id`
-                                       IN (" . implode(",", $_SESSION['glpigroups']) . "))";
-            $condition .= " OR (`$teamtable`.`itemtype` = 'Group'
-                                      AND `$teamtable`.`items_id`
-                                          IN (" . implode(",", $_SESSION['glpigroups']) . "))";
-         }
-         $condition .= ") ";
-      }
+      //get and clean criteria
+      $criteria = self::getVisibilityCriteria();
+      unset($criteria['LEFT JOIN']);
+      $criteria['FROM'] = self::getTable();
+
+      $it = new \DBmysqlIterator(null);
+      $it->execute($criteria);
+      $sql = $it->getSql();
+      $sql = preg_replace('/.*WHERE /', '', $sql);
 
-      return $condition;
+      return $sql;
    }
 
+   /**
+    * Return visibility joins to add to DBIterator parameters
+    *
+    * @since 9.4
+    *
+    * @param boolean $forceall force all joins (false by default)
+    *
+    * @return array
+    */
+   static public function getVisibilityCriteria($forceall = false) {
+      global $CFG_GLPI;
+
+      if (Session::haveRight('project', self::READALL)) {
+         return [
+            'LEFT JOIN' => [],
+            'WHERE' => [],
+         ];
+      }
+
+      $join = [];
+      $where = [];
+
+      $join['glpi_projectteams'] = [
+         'ON' => [
+            'glpi_projectteams'  => 'projects_id',
+            'glpi_projects'      => 'id'
+         ]
+      ];
+
+      $teamtable = 'glpi_projectteams';
+      $where['OR'] = [
```

Comment:
It is dangerous to put it in a `OR` key directly as it will be merged with other conditions that also have this `OR` key.

Reply:
```php
$where[] = ['OR' => [...]];
 ```
?

Reply:
Yes, it will be more secure.
I was wondering if, for this kind of methods, the whole conditions could be put as an array item, like this:
```php
      $criteria = [
         'LEFT JOIN' => $join,
         'WHERE'     => [$where]
      ];
       return $criteria;
```
Doing this would prevent them to be overrided by accident.

Reply:
I was also wondering if the JOIN should be secured, for example by adding aliases, as they also could be overrided when merging with the initial request criteria.

Reply:
I'm affraid this would imply too much changes on the current iterator